### PR TITLE
fix(analyzers/diagnostics): handle invalid fields in TOML

### PR DESCRIPTION
**Pull request by burntcarrot**
Adds a couple of fixes for handling TOML validation errors - including handling invalid/extra fields present in TOML files.

Screencast: https://asciinema.org/a/kqnm8farRPoKi7AneSzTAR6QV

### CODE REVIEW

File: analyzers/validator/utils.go [hunk 1]
```diff
@@ -5,12 +5,16 @@ import (
...
 	"path"
 	"reflect"
+	"regexp"
 	"strings"
 
 	validate "github.com/go-playground/validator/v10"
 	"github.com/pelletier/go-toml/v2"
 )
 
+// codeframeFieldRegexp is a regular expression used for matching a TOML field present in the codeframe generated by go-toml.
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:15]:
```suggestion
// codeframeFieldRegexp is a regular expression used for matching a TOML field present in the codeframe generated by go-toml's DecodeError.
// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#DecodeError
```

File: analyzers/validator/utils.go [hunk 2]
```diff
@@ -32,6 +36,38 @@ func getMissingRequiredFields(err error, config interface{}) []string {
...
+		}
+
+		// Construct a validation failure.
+		validationError := ValidationFailure{
+			File:   filePath,
+			Errors: errors,
+		}
+
+		return &validationError
+	}
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:69]: Please move this to a separate function.

File: analyzers/validator/utils.go [hunk 3]
```diff
@@ -32,6 +36,38 @@ func getMissingRequiredFields(err error, config interface{}) []string {
...
+
 	// Get the DecodeError exported by go-toml
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:71]: Please move this `DecodeError` handler to a separate function as well.

File: analyzers/validator/utils.go [hunk 4]
```diff
@@ -101,3 +137,44 @@ func handleTOMLDecodeErrors(err error, filePath string) *ValidationFailure {
 	}
 	return &validationError
 }
+
+// parseCodeframe parses the codeframe generated by go-toml, and extracts field names which are associated with the error.
+func parseCodeframe(codeframe string) []string {
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:142]: Please rename this function to something like `parseTOMLErrorCodeFrame` in order to give a context as to parsing a code frame reported as an error by the toml library.

**Review by siddhant-deepsource — CHANGES_REQUESTED** [review]
Looks good. Please address the comments. Thanks!

File: analyzers/diagnostics/diagnostics.go [hunk 5]
```diff
@@ -209,6 +209,11 @@ func prepareCodeFrame(lineNum int, lines []string) string {
 
 // checkField checks if the line contains the field.
 func checkField(line, field string) bool {
+	// Sometimes validation errors have a blank field name. In that case, checkField should return false.
```
[the remark this answers is not in the document]
Comment (burntcarrot) [reply analyzers/diagnostics/diagnostics.go:212]: If the field is empty in the validation error itself, `checkField` returns false, which causes the diagnostic to not get printed on the console.

I couldn't find a case where blank field names have been reported, but I've added it as a check in order to ensure that it doesn't break abruptly.

File: analyzers/validator/utils.go [hunk 6]
```diff
@@ -30,37 +33,44 @@ func getMissingRequiredFields(err error, config interface{}) []string {
...
-	 * a validation failure response
-	 * ================================================= */
-
+	// Handle DecodeErr.
 	errorMessage := decodeErr.Error()
+	validationError := handleDecodeErr(errorMessage, filePath)
+	return &validationError
+}
+
+// handleDecodeErr extracts data about the decoding failure and returns a validation failure response.
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:70]:
```suggestion
// handleDecodeErr extracts the required data from the decoding errors of the TOML of type other than DecodeError and StrictMissingError and returns a validation failure response.
```

File: analyzers/validator/utils.go [hunk 7]
```diff
@@ -30,37 +33,44 @@ func getMissingRequiredFields(err error, config interface{}) []string {
...
+	// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#example-Decoder.DisallowUnknownFields
+
+	var strictMissingError *toml.StrictMissingError
+	if errors.As(err, &strictMissingError) {
+		validationError := handleStrictMissingError(err, strictMissingError, filePath)
+
+		return &validationError
+	}
 
 	// Get the DecodeError exported by go-toml
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:47]:
```suggestion
	// Check for the DecodeError exported by go-toml.
```

File: analyzers/validator/utils.go [hunk 8]
```diff
@@ -3,14 +3,17 @@ package validator
...
-	"path"
 	"reflect"
+	"regexp"
 	"strings"
 
 	validate "github.com/go-playground/validator/v10"
 	"github.com/pelletier/go-toml/v2"
 )
 
+// codeframeFieldRegexp is a regular expression used for matching a TOML field present in the codeframe generated by go-toml's StrictMissingError.
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:14]:
```suggestion
/*
codeframeFieldRegexp is a regular expression used for matching a TOML field present in the codeframe generated by go-toml's StrictMissingError.
Ref: The example here: https://pkg.go.dev/github.com/pelletier/go-toml/v2#example-Decoder.DisallowUnknownFields
The output looks like this:
strict mode: fields in the document are missing in the target struct
2| key1 = "value1"
3| key2 = "value2"
 | ~~~~ missing field
4| key3 = "value3"
*/
```

**Review by siddhant-deepsource — APPROVED** [review]
Looks good to me. This improves the config validation for the user thus adding to the improved user experience. Thanks!

File: analyzers/diagnostics/diagnostics.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (siddhant-deepsource) [reply analyzers/diagnostics/diagnostics.go:212]: So, in this case what would the user see on running the command? Will they just see `Invalid fields found`?

File: analyzers/validator/utils.go [hunk 9]
```diff
@@ -99,5 +118,75 @@ func handleTOMLDecodeErrors(err error, filePath string) *ValidationFailure {
 			},
 		},
 	}
-	return &validationError
+
+	return validationError
+}
+
+// handleStrictMissingError handles StrictMissingError reported by go-toml, and returns a validation failure response.
+func handleStrictMissingError(err error, strictMissingError *toml.StrictMissingError, filePath string) ValidationFailure {
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:126]:
```suggestion
func handleStrictMissingError(strictMissingError *toml.StrictMissingError, filePath string) ValidationFailure {
```

File: analyzers/diagnostics/diagnostics.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (burntcarrot) [reply analyzers/diagnostics/diagnostics.go:212]: Ideally this shouldn't occur in any case, but I'll try simulating a edge case somehow locally. I'll let you know about the results.

[the remark this answers is not in the document]
Comment (siddhant-deepsource) [reply analyzers/diagnostics/diagnostics.go:212]: Okay. Resolving for now.

File: analyzers/validator/utils.go (hunk 9 above)
Comment (burntcarrot) [reply analyzers/validator/utils.go]: Nice catch!

Comment (siddhant-deepsource) [reply analyzers/validator/utils.go]: Thanks to the Go Analyzer! 😅

File: analyzers/validator/utils.go [hunk 10]
```diff
@@ -30,37 +42,44 @@ func getMissingRequiredFields(err error, config interface{}) []string {
...
 // Handle decoding errors reported by go-toml
 func handleTOMLDecodeErrors(err error, filePath string) *ValidationFailure {
-	var usefulResponse, expectedType, receivedType, fieldName, decodeErrorMessage string
+	// We check for go-toml's StrictMissingError, as the decoder returns a StrictMissingError when there exists a mismatch between the destination and the extracted fields.
+	// go-toml doesn't return the field name associated with the error, however, StrictMissingError returns a human-readable codeframe, which can be used to extract the error-causing fields.
+	// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#example-Decoder.DisallowUnknownFields
+
+	var strictMissingError *toml.StrictMissingError
+	if errors.As(err, &strictMissingError) {
+		validationError := handleStrictMissingError(err, strictMissingError, filePath)
```
Comment (siddhant-deepsource) [inline analyzers/validator/utils.go:51]:
```suggestion
		validationError := handleStrictMissingError(strictMissingError, filePath)
```

**Review by siddhant-deepsource — APPROVED** [review]

**Comment by siddhant-deepsource** [comment]
@burntcarrot The tests are fixed. Please take a look and do a manual sanity test on your end in order to ensure diagnostics work on the edge cases that we have addressed in this PR. Once done, this is good to go. Thanks!

**Comment by burntcarrot** [comment]
@siddhant-deepsource Did the checks on my end. Works for every test case.
